File: 2022/best-time-to-buy-and-sell-stock/Solution.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length <= 1)
            return 0;
        int buy = prices[0];
        int sell = 0;
        int profit = 0;
        for(int i : prices) {
            if (profit < i - buy)
                profit = i - buy;
            if (i < buy)
                buy = i;
        }
        return profit;
    }
}

File: 2022/best-time-to-buy-and-sell-stock/Solution.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if ( prices.size() <= 1) {
            return 0;
        }
        // [7,1,5,3,6,4]
        // buy = 1
        // cur = 6
        // profit = 5
        int profit = 0;
        int buy = prices[0];
        for(int i = 1; i < prices.size(); ++i) {
            const int cur = prices[i];

           // check if cur profit is bigger from last time we bought
           if (profit < (cur - buy))
                profit = cur - buy;

            // check if the prices is lower
            if (cur < buy)
                buy = cur;
        }
        return profit;
    }
};
